apps/worker/src/monitor/tx.status.monitor.service.ts

```diff
@@ -0,0 +1,147 @@
+import { ITxMonitorJob } from '../interfaces/status-monitor.interface';
+import { QueueConstants } from '../../../../libs/common/src';
+import { SECONDS_PER_BLOCK } from '../../../../libs/common/src/constants';
+import { BlockchainConstants } from '../../../../libs/common/src/blockchain/blockchain-constants';
+
+@Injectable()
+@Processor(QueueConstants.TRANSACTION_RECEIPT_QUEUE_NAME, {
+  concurrency: 2,
+})
+export class TxStatusMonitoringService extends WorkerHost implements OnApplicationBootstrap, OnModuleDestroy {
+  private logger: Logger;
+
+  constructor(
+    @InjectRedis() private cacheManager: Redis,
+    @InjectQueue(QueueConstants.TRANSACTION_RECEIPT_QUEUE_NAME) private txReceiptQueue,
+    @InjectQueue(QueueConstants.PUBLISH_QUEUE_NAME) private publishQueue: Queue,
+    private blockchainService: BlockchainService,
+    private configService: ConfigService,
+    private eventEmitter: EventEmitter2,
+  ) {
+    super();
+    this.logger = new Logger(this.constructor.name);
+  }
+
+  public async onApplicationBootstrap() {
+    this.logger.debug('Starting publishing service');
+  }
+
+  public onModuleDestroy() {
+    try {
+      this.logger.debug('Shutting down publishing service');
+    } catch (e) {
+      // 🐂 //
+    }
+  }
+
+  async process(job: Job<ITxMonitorJob, any, string>): Promise<any> {
+    this.logger.log(`Monitoring job ${job.id} of type ${job.name}`);
+    try {
+      const numberBlocksToParse = BlockchainConstants.NUMBER_BLOCKS_TO_CRAWL;
+      const txCapacityEpoch = job.data.epoch;
+      const previousKnownBlockNumber = (await this.blockchainService.getBlock(job.data.lastFinalizedBlockHash)).block.header.number.toBigInt();
+      const currentFinalizedBlockNumber = await this.blockchainService.getLatestFinalizedBlockNumber();
+      const blockList: bigint[] = [];
+      for (let i = previousKnownBlockNumber; i <= currentFinalizedBlockNumber && i < previousKnownBlockNumber + numberBlocksToParse; i += 1n) {
+        blockList.push(i);
+      }
+      const txBlockHash = await this.crawlBlockList(job.data.txHash, txCapacityEpoch, blockList);
+
+      if (txBlockHash) {
+        this.logger.verbose(`Successfully completed job ${job.id}`);
+        return { success: true };
+      }
+
+      // handle failure to find tx in block list after
+      if (!txBlockHash && job.attemptsMade >= (job.opts.attempts ?? 3)) {
+        this.logger.error(`Job failed max attempts ${job.attemptsMade}, enqueueing to publish queue`);
+        await this.publishQueue.removeRepeatableByKey(job.data.referencePublishJob.id);
+        const publishJob = {
+          id: job.data.referencePublishJob.id,
+          schemaId: job.data.referencePublishJob.schemaId,
+          data: job.data.referencePublishJob.data,
+        };
+
+        const delay = 1 * MILLISECONDS_PER_SECOND * SECONDS_PER_BLOCK;
+        await this.publishQueue.add(QueueConstants.PUBLISH_QUEUE_NAME, publishJob, { delay });
```

Comment:
I think we need to use the same jobId as the one used in previous publishQueue and if the job already exist we should remove it before enqueue

Reply:
cool